src/sentry/sentry_metrics/indexer/limiters/cardinality.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import dataclasses
+from collections import defaultdict
+from typing import Any, Generic, Mapping, MutableMapping, Optional, Sequence, TypedDict, TypeVar
+
+from sentry import options
+from sentry.ratelimits.cardinality import (
+    CardinalityLimiter,
+    GrantedQuota,
+    Quota,
+    RedisCardinalityLimiter,
+    RequestedQuota,
+    Timestamp,
+)
+from sentry.sentry_metrics.configuration import MetricsIngestConfiguration, UseCaseKey
+from sentry.utils import metrics
+from sentry.utils.hashlib import hash_values
+
+OrgId = int
+
+TMessageKey = TypeVar("TMessageKey")
+
+
+@dataclasses.dataclass(frozen=True)
+class CardinalityLimiterState(Generic[TMessageKey]):
+    _cardinality_limiter: CardinalityLimiter
+    _use_case_id: UseCaseKey
+    _grants: Sequence[GrantedQuota]
+    _timestamp: Timestamp
+    keys_to_remove: Sequence[TMessageKey]
+
+
+def _build_quota_key(namespace: str, org_id: Optional[OrgId]) -> str:
+    if org_id is not None:
+        return f"metrics-indexer-cardinality-{namespace}-org-{org_id}"
+    else:
+        return f"metrics-indexer-cardinality-{namespace}-global"
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quotas")
+def _construct_quotas(use_case_id: UseCaseKey) -> Sequence[Quota]:
+    """
+    Construct write limit's quotas based on current sentry options.
+
+    This value can potentially cached globally as long as it is invalidated
+    when sentry.options are.
+    """
+    if use_case_id == UseCaseKey.PERFORMANCE:
+        return [
+            Quota(**args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+        ]
+    elif use_case_id == UseCaseKey.RELEASE_HEALTH:
+        return [
+            Quota(**args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+class InboundMessage(TypedDict):
+    # Note: This is only the subset of fields we access in this file.
+    org_id: int
+    name: str
+    type: str
+    tags: Mapping[str, str]
+
+
+class TimeseriesCardinalityLimiter:
+    def __init__(self, namespace: str, rate_limiter: CardinalityLimiter) -> None:
+        self.namespace = namespace
+        self.rate_limiter: CardinalityLimiter = rate_limiter
+
+    def check_cardinality_limits(
+        self, use_case_id: UseCaseKey, messages: Mapping[TMessageKey, InboundMessage]
+    ) -> CardinalityLimiterState[TMessageKey]:
+        request_hashes = defaultdict(set)
+        keys_to_remove = {}
+        for key, message in messages.items():
+            org_id = message["org_id"]
+            message_hash = int(
+                hash_values(
+                    [
+                        message["name"],
+                        message["type"],
```

Comment:
I don't think we need `type` in the hash. If the same time series show up in two types (which are different tables in clickhouse) we can count only 1.

Reply:
that's fair, `type` is not a very large dimension